core/src/main/kotlin/com/strumenta/kolasu/model/observable/ReferenceToNodeObserver.kt

```diff
@@ -0,0 +1,44 @@
+package com.strumenta.kolasu.model.observable
+
+import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.model.PossiblyNamed
+import com.strumenta.kolasu.model.ReferenceChangeNotification
+import io.reactivex.rxjava3.core.Observer
+import io.reactivex.rxjava3.disposables.Disposable
+
+class ReferenceToNodeObserver<N : PossiblyNamed>(val container: Node, val referenceName: String) :
```

Comment:
I don't understand the purpose of this class very well, but I have the sensation this should not be an observer. An observer acts as a door out of the Rx framework, and calling `onNext` from an observer directly usually is a code smell. The idiomatic way to do this would be to use an operator such as [map](https://github.com/ReactiveX/RxJava/wiki/Transforming-Observables#map) if the purpose is to transform events from one type to the other. E.g. something like

```
ref.changes.map(...).subscribe(node.notifications)
```

Reply:
The idea is to use it to propagate the changes on the references to the node, in a convenient way. For example:
```
data class NodeWithReference(val ref: ReferenceByName<NamedNode>, val id: Int) : Node() {
    init {
        ref.subscribe(ReferenceToNodeObserver(this, "ref"))
    }
}
```
This code should either written manually or injected by the compiler plugin through a cumbersome API, so the simpler is the call, the better

Reply:
I understand I should change the example in ObservableNodeTest to something like:

```
data class NodeWithReference(val ref: ReferenceByName<NamedNode>, val id: Int) : Node() {
    init {
        //ref.subscribe(ReferenceToNodeObserver(this, "ref"))
        ref.changes.map(...).subscribe(this.changes)
    }
}
```

But it is not clear to me what should I specify where the dots are. Could you help me on this?

Reply:
Something like 

```
ref.changes.map { referenceNotification -> 
    return if (referenceNotification.oldValue != null) {
        ReferencedToRemoved(referenceNotification.oldValue, referenceName, container)
    } else if (referenceNotification.newValue != null) {
        ReferencedToAdded(referenceNotification.newValue, referenceName, container)
    } else ??? // I have no idea if this is possible/what it would represent, if the idea is that this should be dropped then the mergeMap operator is more adequate
}
```

Reply:
What we want to do is this:
* Always notify the owner of the reference about the change
* Notify the previous target of the reference (if it was not null) about the change
* Notify the next target of the reference (if it is not null) about the change       

For the first point I think we can write:
```
        ref.changes.map { referenceNotification ->
            ReferenceSet(
                container,
                referenceName,
                referenceNotification.oldValue as Node?,
                referenceNotification.newValue as Node?
            ) as NodeNotification<in Node>
        }.subscribe(this.changes)
```

However I do not know how to specify the other two cases, as in that case the thing that should receive the notification depend dynamically from the past and future value of the reference target.

You can see that in `ReferenceToNodeObserver`:

```
    override fun onNext(referenceNotification: ReferenceChangeNotification<N>) {
        container.changes.onNext(
            ReferenceSet(
                container,
                referenceName,
                referenceNotification.oldValue as Node?,
                referenceNotification.newValue as Node?
            )
        )
        if (referenceNotification.oldValue != null) {
            referenceNotification.oldValue.changes.onNext(
                ReferencedToRemoved(referenceNotification.oldValue, referenceName, container)
            )
        }
        if (referenceNotification.newValue != null) {
            referenceNotification.newValue.changes.onNext(
                ReferencedToAdded(referenceNotification.newValue, referenceName, container)
            )
        }
    }
```

Is this possible to do using map or with another call?            

I tried with this:

```
        ref.changes.mapOptional { referenceNotification ->
            if (referenceNotification.oldValue != null) {
                Optional.of(ReferencedToRemoved(referenceNotification.oldValue!!, referenceName, container))
            } else {
                Optional.empty()
            } as Optional<ReferencedToRemoved<in Node>>
        }.subscribe { referenceNotification ->
            (referenceNotification.node as Node).changes
        }
        ref.changes.mapOptional { referenceNotification ->
            if (referenceNotification.newValue != null) {
                Optional.of(ReferencedToAdded(referenceNotification.newValue!!, referenceName, container))
            } else {
                Optional.empty()
            } as Optional<ReferencedToAdded<in Node>>
        }.subscribe { referenceNotification ->
            (referenceNotification.node as Node).changes
        }
```

But it does not seem to work

Reply:
Thank you, your solution looks much better. I fixed the issue with the IR plugin. Basically it needs to generate for each AST class and each references it has this code:

```
init {
       ref.setContainer(this, "ref name")
}
```